src/map/status.c

```diff
@@ -12325,6 +12291,151 @@ bool status_readdb_scconfig(char* fields[], int columns, int current) {
+ * @param n       Ordinal number of the entry, to be displayed in case of
+ *                validation errors.
+ * @param *source Source of the entry (file name), to be displayed in case of
+ *                validation errors.
+ * @return # of the validated entry, or 0 in case of failure.
+ */
+int status_readdb_refine_libconfig_sub(config_setting_t *r, int n, const char *source) {
+	config_setting_t *t = NULL;
+	int i, type, bonus_per_level, random_bonus, random_bonus_start_level;
+	
+	if( !status->refinedb_lookup_const(r, "wLvl", &type) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Invalid or missing wLvl in \"%s\", entry #%d, skipping.\n", source, n);
+		return 0;
+	} else if ( type < REFINE_TYPE_ARMOR || type >= REFINE_TYPE_MAX ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Invalid Type for entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	} else if( !status->refinedb_lookup_const(r, "StatsPerLevel", &bonus_per_level) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing StatsPerLevel in entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	} else if( !status->refinedb_lookup_const(r, "RandomBonusStartLevel", &random_bonus_start_level) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing RandomBonusStartLevel in entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	} else if ( !status->refinedb_lookup_const(r, "RandomBonusValue", &random_bonus) ) {
+		ShowWarning("status_readdb_refine_libconfig_sub: Missing RandomBonusValue in entry #%d of \"%s\", skipping.\n", n, source);
+		return 0;
+	}
+	
+	if ((t=libconfig->setting_get_member(r, "Rate")) && config_setting_is_list(t)) {
+		config_setting_t *tt = NULL;
+		bool duplicate[MAX_REFINE];
+		
+		memset(&duplicate,0,sizeof(duplicate));
+		
+		for(i=0; i < MAX_REFINE; i++) {
+			tt = libconfig->setting_get_elem(t, i);
+			
+			if (!tt)
+				break;
+			else if (!config_setting_is_group(tt))
+				continue;
+			
+			int level=0, chance=0;
+			
+			if (!libconfig->setting_lookup_int(tt, "Level", &level) || level <= 0 || level > MAX_REFINE) {
+				ShowError("status_readdb_refine_libconfig_sub: Invalid 'Level' configuration %d in entry #%d of \"%s\".\n", level, n, source);
+				return 0;
+			}  else if (!libconfig->setting_lookup_int(tt, "Chance", &chance)) {
+				ShowWarning("status_readdb_refine_libconfig_sub: Missing 'Chance' configuration in entry #%d of \"%s\", defaulting to 0.\n", n, source);
+				status->dbs->refine_info[type].chance[i] = 0;
+			}
+			
+			level -= 1;
+			
+			if( duplicate[level] ) {
+				ShowWarning("status_readdb_refine_libconfig_sub: duplicate entry of Level %d in Type #%d of \"%s\".\n", level, type, source);
+			} else duplicate[level] = true;
+			
+			if ( chance ) status->dbs->refine_info[type].chance[level] = chance;
+			else status->dbs->refine_info[type].chance[level] = 0;
+		}
+		for(i=0; i < MAX_REFINE; i++) {
+			int bonus=0;
+			tt = libconfig->setting_get_elem(t, i);
```

Comment:
This loop assumes that the levels are in order (while the above does otherwise). Instead of calling once again `libconfig->setting_get_elem(t, i)`, you could just cache the Bonus value to a temporary array in the loop above, and here simply loop through that array